### PR TITLE
Add machine codes and their handling in gatekeeper

### PR DESCRIPTION
**Pull request by rwrzesien**
Partially resolves https://github.com/golemfactory/concent/issues/131.

@cameel According to your comment in https://github.com/golemfactory/concent/pull/133, this is part related to gatekeeper, rebased on latest master. We can discuss here.

### CODE REVIEW

File: concent_api/gatekeeper/views.py [hunk 1]
```diff
@@ -78,36 +86,65 @@ def download(request):
...
+            "Token in the 'Authorization' header is not a valid Golem message.",
+            ErrorCode.HEADER_AUTHORIZATION_TOKEN_NOT_VALID_MESSAGE.value,
+            path_to_file,
+        )
 
     if loaded_golem_message is None:
-        return gatekeeper_access_denied_response('Undefined golem message.', path_to_file)
+        return gatekeeper_access_denied_response(
+            'Undefined golem message.',
+            ErrorCode.MESSAGE.value,
```
Comment (cameel) [inline concent_api/gatekeeper/views.py]: This looks more like `MESSAGE_UNABLE_TO_DESERIALIZE` or `HEADER_AUTHORIZATION_TOKEN_INVALID_MESSAGE`.

But can this really happen? I think that now `load()` is never supposed to return `None` and we could replace it with an `assert`.

File: concent_api/gatekeeper/views.py [hunk 2]
```diff
@@ -35,9 +36,13 @@
 @require_POST
 def upload(request):
     logger.debug("Upload request received.")
+
     if request.content_type in ['multipart/form-data', '', None] or request.content_type.isspace():
         logger.info(f'Unsupported content type: {request.content_type}')
-        return gatekeeper_access_denied_response('Unsupported content type.')
+        return gatekeeper_access_denied_response(
+            'Unsupported content type.',
+            ErrorCode.HEADER_CONTENT_TYPE_NOT_SUPPORTED.value,
```
Comment (cameel) [inline concent_api/gatekeeper/views.py]: Let's use enum items, not names or values. By taking the value immediately we lose type information and only make expressions longer. Enum items should be converted to values only when they need to be serialized, i.e. in  `gatekeeper_access_denied_response()`.

Comment (cameel) [inline concent_api/gatekeeper/views.py]: That's not part of this pull request but maybe we could change it anyway: this should be HTTP 415.

File: concent_api/gatekeeper/views.py [hunk 3]
```diff
@@ -78,36 +86,65 @@ def download(request):
...
+            path_to_file,
+        )
 
     try:
         decoded_auth_header_content = b64decode(token, validate = True)
     except binascii.Error:
-        return gatekeeper_access_denied_response("Unable to decode token in the 'Authorization' header.", path_to_file)
+        return gatekeeper_access_denied_response(
+            "Unable to decode token in the 'Authorization' header.",
+            ErrorCode.HEADER_AUTHORIZATION_UNABLE_TO_DECODE.value,
```
Comment (cameel) [inline concent_api/gatekeeper/views.py]: Maybe `HEADER_AUTHORIZATION_NOT_BASE64`?

File: concent_api/gatekeeper/views.py [hunk 4]
```diff
@@ -78,36 +86,65 @@ def download(request):
...
+            path_to_file,
+        )
 
     try:
         loaded_golem_message = load(decoded_auth_header_content, settings.CONCENT_PRIVATE_KEY, settings.CONCENT_PUBLIC_KEY, check_time = False)
     except MessageError:
-        return gatekeeper_access_denied_response("Token in the 'Authorization' header is not a valid Golem message.", path_to_file)
+        return gatekeeper_access_denied_response(
+            "Token in the 'Authorization' header is not a valid Golem message.",
+            ErrorCode.HEADER_AUTHORIZATION_TOKEN_NOT_VALID_MESSAGE.value,
```
Comment (cameel) [inline concent_api/gatekeeper/views.py]: Maybe `HEADER_AUTHORIZATION_TOKEN_INVALID_MESSAGE`?

File: concent_api/utils/constants.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    HEADER_AUTHORIZATION_UNRECOGNIZED_SCHEME                = 'header.authorization.unrecognized_scheme'
+    HEADER_AUTHORIZATION_UNABLE_TO_DECODE                   = 'header.authorization.unable_to_decode'
+    HEADER_CLIENT_PUBLIC_KEY_MISSING                        = 'header.client_public_key.missing'
+    HEADER_CLIENT_PUBLIC_KEY_NOT_BASE64_ENCODED_VALUE       = 'header.client_public_key.not_base64_encoded_value'
+    HEADER_CLIENT_PUBLIC_KEY_WRONG_LENGTH                   = 'header.client_public_key.wrong_length'
+    HEADER_CONTENT_TYPE_MISSING                             = 'header.content_type.missing'
+    HEADER_CONTENT_TYPE_NOT_SUPPORTED                       = 'header.content_type.not_supported'
+    HEADER_OTHER_PARTY_PUBLIC_KEY_MISSING                   = 'header.other_party_public_key.missing'
+    HEADER_OTHER_PARTY_PUBLIC_KEY_NOT_BASE64_ENCODED_VALUE  = 'header.other_party_public_key.not_base64_encoded_value'
+    MESSAGE                                                 = 'message'
```
Comment (cameel) [inline concent_api/utils/constants.py]: I think `MESSAGE` is too generic.

File: concent_api/gatekeeper/views.py [hunk 6]
```diff
@@ -78,36 +86,65 @@ def download(request):
...
+        )
     assert isinstance(loaded_golem_message, Message)
 
     # Check if request header contains Concent-Client-Public-Key:
     if 'HTTP_CONCENT_CLIENT_PUBLIC_KEY' not in request.META:
-        return gatekeeper_access_denied_response('Missing Concent-Client-Public-Key header.', path_to_file, loaded_golem_message.subtask_id)
+        return gatekeeper_access_denied_response(
+            'Missing Concent-Client-Public-Key header.',
+            ErrorCode.HEADER_CLIENT_PUBLIC_KEY_MISSING.value,
+            path_to_file,
```
Comment (cameel) [inline concent_api/gatekeeper/views.py]: This header has been removed. I think this will disappear when you rebase.

File: concent_api/gatekeeper/views.py [hunk 7]
```diff
@@ -124,73 +161,193 @@ def parse_headers(request: WSGIRequest, path_to_file: str) -> Union[FileTransfer
...
+            ErrorCode.MESSAGE_STORAGE_CLUSTER_INVALID_URL.value,
+            path_to_file,
+            loaded_golem_message.subtask_id,
+            client_public_key,
+        )
     if loaded_golem_message.storage_cluster_address != settings.STORAGE_CLUSTER_ADDRESS:
-        return gatekeeper_access_denied_response('Given storage_cluster_address is not defined in settings.', path_to_file, loaded_golem_message.subtask_id, client_public_key)
+        return gatekeeper_access_denied_response(
+            'Given storage_cluster_address is not defined in settings.',
+            ErrorCode.MESSAGE_STORAGE_CLUSTER_UNKNOWN.value,
```
Comment (cameel) [inline concent_api/gatekeeper/views.py]: I'd reword the message and the error
``` python
'This token does not allow file transfers to/from the cluster you're trying to access.',
ErrorCode.MESSAGE_STORAGE_WRONG_CLUSTER.value,
```

File: concent_api/gatekeeper/views.py [hunk 8]
```diff
@@ -124,73 +161,193 @@ def parse_headers(request: WSGIRequest, path_to_file: str) -> Union[FileTransfer
...
+            path_to_file,
+            loaded_golem_message.subtask_id,
+            client_public_key,
+        )
 
     # -CLIENT_PUBLIC_KEY
     if not isinstance(loaded_golem_message.authorized_client_public_key, bytes):
-        return gatekeeper_access_denied_response('Wrong type of authorized_client_public_key variable.', path_to_file, loaded_golem_message.subtask_id, client_public_key)
+        return gatekeeper_access_denied_response(
+            'Wrong type of authorized_client_public_key variable.',
```
Comment (cameel) [inline concent_api/gatekeeper/views.py]: I did not point this out in earlier reviews, but it's a bit confusing to call it `variable`. I know it was here already when you started working on it but could you change it to `field` in all error messages?

File: concent_api/utils/constants.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
+import enum
+
+
+class ErrorCode(enum.Enum):
+    BODY_NOT_EMPTY                                          = 'body.not_empty'
```
Comment (cameel) [inline concent_api/utils/constants.py]: Maybe `REQUEST_BODY_NOT_EMPTY`?

File: concent_api/gatekeeper/views.py [hunk 10]
```diff
@@ -124,73 +161,193 @@ def parse_headers(request: WSGIRequest, path_to_file: str) -> Union[FileTransfer
...
+            path_to_file,
+            loaded_golem_message.subtask_id,
+            client_public_key,
+        )
     client_public_key_base64 = (b64encode(loaded_golem_message.authorized_client_public_key)).decode('ascii')
     if client_public_key != client_public_key_base64:
-        return gatekeeper_access_denied_response('authorized_client_public_key is different then Concent-Client-Public-Key header.', path_to_file, loaded_golem_message.subtask_id, client_public_key)
+        return gatekeeper_access_denied_response(
+            'authorized_client_public_key is different then Concent-Client-Public-Key header.',
+            ErrorCode.MESSAGE_AUTHORIZED_CLIENT_PUBLIC_KEY_INVALID.value,
```
Comment (cameel) [inline concent_api/gatekeeper/views.py]: I'd reword the message and the error
``` python
'You're not authorized to use this token.',
ErrorCode.MESSAGE_AUTHORIZED_CLIENT_PUBLIC_KEY_UNAUTHORIZED_CLIENT.value,
```

File: concent_api/gatekeeper/views.py [hunk 11]
```diff
@@ -124,73 +161,193 @@ def parse_headers(request: WSGIRequest, path_to_file: str) -> Union[FileTransfer
...
+            path_to_file,
+            loaded_golem_message.subtask_id,
+            client_public_key,
+        )
 
     # -OPERATION
     if request.method == 'POST' and loaded_golem_message.operation != 'upload':
-        return gatekeeper_access_denied_response('Wrong operation variable for this request method.', path_to_file, loaded_golem_message.subtask_id, client_public_key)
+        return gatekeeper_access_denied_response(
+            'Wrong operation value for this request method.',
```
Comment (cameel) [inline concent_api/gatekeeper/views.py]: `Wrong operation value for this request method` -> `Upload requests must use POST method`.

File: concent_api/gatekeeper/views.py [hunk 12]
```diff
@@ -124,73 +161,193 @@ def parse_headers(request: WSGIRequest, path_to_file: str) -> Union[FileTransfer
...
+            ErrorCode.MESSAGE_OPERATION_INVALID.value,
+            path_to_file,
+            loaded_golem_message.subtask_id,
+            client_public_key,
+        )
     if request.method in ['GET', 'HEAD'] and loaded_golem_message.operation != 'download':
-        return gatekeeper_access_denied_response('Wrong operation variable for this request method.', path_to_file, loaded_golem_message.subtask_id, client_public_key)
+        return gatekeeper_access_denied_response(
+            'Wrong operation value for this request method.',
+            ErrorCode.MESSAGE_OPERATION_INVALID.value,
```
Comment (cameel) [inline concent_api/gatekeeper/views.py]: `Wrong operation value for this request method` -> `Download requests must use GET or HEAD method`.

File: concent_api/gatekeeper/views.py [hunk 13]
```diff
@@ -124,73 +161,193 @@ def parse_headers(request: WSGIRequest, path_to_file: str) -> Union[FileTransfer
...
+            path_to_file,
+            loaded_golem_message.subtask_id,
+            client_public_key
+        )
     transfer_token_paths_to_files = [file["path"] for file in loaded_golem_message.files]
     if len(transfer_token_paths_to_files) != len(set(transfer_token_paths_to_files)):
-        return gatekeeper_access_denied_response('File paths in the token must be unique', path_to_file, loaded_golem_message.subtask_id, client_public_key)
+        return gatekeeper_access_denied_response(
+            'File paths in the token must be unique',
+            ErrorCode.MESSAGE_FILES_PATH_NOT_UNIQUE.value,
```
Comment (cameel) [inline concent_api/gatekeeper/views.py]: `MESSAGE_FILES_PATH_NOT_UNIQUE` -> `MESSAGE_FILES_PATHS_NOT_UNIQUE`

File: concent_api/gatekeeper/views.py [hunk 14]
```diff
@@ -124,73 +161,193 @@ def parse_headers(request: WSGIRequest, path_to_file: str) -> Union[FileTransfer
...
+            path_to_file,
+            loaded_golem_message.subtask_id,
+            client_public_key,
+        )
     file_checksums = [tuple(file["checksum"].split(":")) for file in loaded_golem_message.files]
     if not set(checksum_type for checksum_type, checksum in file_checksums).issubset(set(HashingAlgorithm._value2member_map_)):  # type: ignore
-        return gatekeeper_access_denied_response("Unrecognized checksum type.", path_to_file, loaded_golem_message.subtask_id, client_public_key)
+        return gatekeeper_access_denied_response(
+            "Unrecognized checksum type.",
+            ErrorCode.MESSAGE_FILES_CHECKSUM_UNRECOGNIZED_TYPE.value,
```
Comment (cameel) [inline concent_api/gatekeeper/views.py]: `Unrecognized checksum type` -> `One of the checksums is from an unsupported hashing algorithm.`
`MESSAGE_FILES_CHECKSUM_UNRECOGNIZED_TYPE` -> `MESSAGE_FILES_CHECKSUM_INVALID_ALGORITHM`

File: concent_api/gatekeeper/views.py [hunk 15]
```diff
@@ -124,73 +161,193 @@ def parse_headers(request: WSGIRequest, path_to_file: str) -> Union[FileTransfer
...
+            loaded_golem_message.subtask_id,
+            client_public_key,
+        )
 
     file_sizes = [file["size"] for file in loaded_golem_message.files]
     if any((file_size is None) for file_size in file_sizes):
-        return gatekeeper_access_denied_response("'size' must be an integer.", path_to_file, loaded_golem_message.subtask_id, client_public_key)
+        return gatekeeper_access_denied_response(
+            "'size' must be an integer.",
+            ErrorCode.MESSAGE_FILES_SIZE_WRONG_TYPE.value,
```
Comment (cameel) [inline concent_api/gatekeeper/views.py]: Maybe `MESSAGE_FILES_SIZE_EMPTY`? We discern between "wrong type" and "empty" for the checksum field. They should be handled the same way. So either both always use "wrong type" or both discern between these two situations.

File: concent_api/gatekeeper/views.py [hunk 16]
```diff
@@ -124,73 +161,193 @@ def parse_headers(request: WSGIRequest, path_to_file: str) -> Union[FileTransfer
...
     matching_files = [file for file in loaded_golem_message.files if path_to_file == file['path']]
 
     if len(matching_files) == 1:
         return matching_files[0]
     else:
         assert len(matching_files) == 0
-        return gatekeeper_access_denied_response('Path to specified file is not listed in files variable.', path_to_file, loaded_golem_message.subtask_id, client_public_key)
+        return gatekeeper_access_denied_response(
+            'Path to specified file is not listed in files variable.',
+            ErrorCode.MESSAGE_FILES_PATH_NOT_LISTED_IN_FILES.value,
```
Comment (cameel) [inline concent_api/gatekeeper/views.py]: `Path to specified file is not listed in files variable.` -> `Your token does not authorize you to transfer the requested file.`

File: concent_api/utils/constants.py [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+    MESSAGE_TOKEN_EXPIRATION_DEADLINE_WRONG_TYPE            = 'message.token_expiration_deadline_wrong_type'
+    MESSAGE_VALUE_NOT_INTEGER                               = 'message.value_not_integer'
+    MESSAGE_UNABLE_TO_DESERIALIZE                           = 'message.unable_to_deserialize'
+    MESSAGE_UNEXPECTED                                      = 'message.unexpected'
+    MESSAGE_UNKNOWN                                         = 'message.unknown'
+    MESSAGE_WRONG_FIELDS                                    = 'message.wrong_fields'
+    QUEUE_COMMUNICATION_NOT_STARTED                         = 'queue.communication_not_started'
+    QUEUE_MESSAGE_ALREADY_PROCESSED                         = 'queue.message_already_processed'
+    QUEUE_RESPONSE_ALREADY_SUBMITTED                        = 'queue.response_already_submitted'
+    QUEUE_TIMEOUT                                           = 'queue.timeout'
```
Comment (cameel) [inline concent_api/utils/constants.py]: Not all codes are used in the code. Please leave only the ones that are that in subsequent pull requests it's easy to see which ones you're adding.

It's hard to say if these codes will be a good idea without seeing how they're used first.

File: concent_api/gatekeeper/utils.py [hunk 18]
```diff
@@ -1,9 +1,10 @@
 from django.http import JsonResponse
 
 
-def gatekeeper_access_denied_response(message, path = None, subtask_id = None, client_key = None):
+def gatekeeper_access_denied_response(message, error_message = None, path = None, subtask_id = None, client_key = None):
     data = {
         'message':         message,
+        'error_code':      error_message,
```
[the remark this answers is not in the document]
Comment (rwrzesien) [reply concent_api/gatekeeper/utils.py]: :+1: Just need renaming, will do.

File: concent_api/gatekeeper/views.py (hunk 1 above)
Comment (rwrzesien) [reply concent_api/gatekeeper/views.py]: There is an assert below that line, but I would like not to modify anything in logic with this pull request. We have separate issue for refactoring gatekeeper views. However I will change to `MESSAGE_UNABLE_TO_DESERIALIZE` as it seems more appropriate.

File: concent_api/gatekeeper/views.py (hunk 2 above)
Comment (rwrzesien) [reply concent_api/gatekeeper/views.py]: :+1:

Comment (rwrzesien) [reply concent_api/gatekeeper/views.py]: :+1:

File: concent_api/gatekeeper/views.py (hunk 3 above)
Comment (rwrzesien) [reply concent_api/gatekeeper/views.py]: :+1:

File: concent_api/utils/constants.py (hunk 5 above)
Comment (rwrzesien) [reply concent_api/utils/constants.py]: :+1:

File: concent_api/gatekeeper/views.py (hunk 6 above)
Comment (rwrzesien) [reply concent_api/gatekeeper/views.py]: :+1: Yes, I am currently updating this part in rebase.

File: concent_api/gatekeeper/views.py (hunk 8 above)
Comment (rwrzesien) [reply concent_api/gatekeeper/views.py]: :+1:

File: concent_api/utils/constants.py (hunk 9 above)
Comment (rwrzesien) [reply concent_api/utils/constants.py]: :+1:

File: concent_api/gatekeeper/views.py (hunk 10 above)
Comment (rwrzesien) [reply concent_api/gatekeeper/views.py]: :+1:

File: concent_api/gatekeeper/views.py (hunk 11 above)
Comment (rwrzesien) [reply concent_api/gatekeeper/views.py]: :+1:

File: concent_api/gatekeeper/views.py (hunk 12 above)
Comment (rwrzesien) [reply concent_api/gatekeeper/views.py]: :+1:

File: concent_api/gatekeeper/views.py (hunk 13 above)
Comment (rwrzesien) [reply concent_api/gatekeeper/views.py]: :+1:

File: concent_api/gatekeeper/views.py (hunk 14 above)
Comment (rwrzesien) [reply concent_api/gatekeeper/views.py]: :+1:

File: concent_api/gatekeeper/views.py (hunk 15 above)
Comment (rwrzesien) [reply concent_api/gatekeeper/views.py]: :+1:

File: concent_api/gatekeeper/views.py (hunk 16 above)
Comment (rwrzesien) [reply concent_api/gatekeeper/views.py]: :+1:

File: concent_api/utils/constants.py (hunk 17 above)
Comment (rwrzesien) [reply concent_api/utils/constants.py]: :+1:

**Comment by rwrzesien** [comment]
@cameel Rebased on latest master and all fixes applied. If branch looks ok, lets merge it and I will prepare second part of it covering the rest of the code.

**Comment by cameel** [comment]
Please mention the original issue in your description too: #131. Without it there's no automatic link back from it to this pull request.

File: concent_api/gatekeeper/views.py [hunk 19]
```diff
@@ -120,6 +152,7 @@ def parse_headers(request: WSGIRequest, path_to_file: str) -> Union[FileTransfer
     except (MessageError, TypeError):
```
Comment (cameel) [inline concent_api/gatekeeper/views.py]: Not necessarily a part of this task but it's a small change:

Can we still get `TypeError` here? If not, we should remove it.

File: concent_api/gatekeeper/views.py (hunk 2 above)
Comment (cameel) [reply concent_api/gatekeeper/views.py]: Sorry for adding you extra work but I was wrong here. It can't be HTTP 415. Normally it should, but the `auth_request` module in nginx can only handle HTTP 401. So please change it back. The status code parameter can stay. Just change the number.

Could you also add a comment explaining it so that this does not happen again? It's not obvious and now I fell into this trap myself even though I know about it :)

File: concent_api/gatekeeper/tests/test_gatekeeper_views.py [hunk 20]
```diff
@@ -216,16 +224,17 @@ def test_upload_should_return_401_if_size_of_file_is_wrong(self):
...
     @freeze_time("2018-12-30 11:00:00")
     def test_upload_should_return_401_if_newlines_in_message_checksum(self):
-        invalid_values = [
-            's\nha1:95a0f391c7ad866\n86ab1366bcd519ba5ab3cce89',
-            '\nsha1:95a0f391c7ad86686ab1366bcd519ba5ab3cce89',
-            'sha1:95a0f391c7ad86686ab1366bcd519ba5ab3cce89\n',
-        ]
+        invalid_values_with_expected_error_code = {
+            's\nha1:95a0f391c7ad866\n86ab1366bcd519ba5ab3cce89': ErrorCode.MESSAGE_FILES_CHECKSUM_INVALID_ALGORITHM.value,
+            '\nsha1:95a0f391c7ad86686ab1366bcd519ba5ab3cce89': ErrorCode.MESSAGE_FILES_CHECKSUM_INVALID_ALGORITHM.value,
```
[the remark this answers is not in the document]
Comment (rwrzesien) [reply concent_api/gatekeeper/tests/test_gatekeeper_views.py]: :+1:

File: concent_api/gatekeeper/views.py (hunk 19 above)
Comment (rwrzesien) [reply concent_api/gatekeeper/views.py]: `TypeError` - no, but I think we can get `binascii.Error` from `b64decode`, so I will replace it.

File: concent_api/gatekeeper/views.py (hunk 2 above)
Comment (rwrzesien) [reply concent_api/gatekeeper/views.py]: :+1: done
